src/main/java/com/minecolonies/coremod/colony/requestsystem/requests/AbstractRequest.java

```diff
@@ -0,0 +1,402 @@
+     * A RequestResolver can compare this object however way it sees fit.
+     * <p>
+     * During the resolving process this object is called multiple times. But at least twice.
+     * A cached implementation is preferred.
+     *
+     * @return The object that is actually requested.
+     */
+    @NotNull
+    @Override
+    public R getRequest()
+    {
+        return requested;
+    }
+
+    /**
+     * Returns the result of this request.
+     *
+     * @return The result of this request, or null if it is not available.
+     */
+    @Nullable
+    @Override
+    public R getResult()
+    {
+        return result;
+    }
+
+    /**
+     * Setter for the result of the request.
+     *
+     * @param result The new result of this request.
+     */
+    @Override
+    public void setResult(@NotNull R result)
+    {
+        this.result = result;
+    }
+
+    /**
+     * Method used to check if the result has been set.
+     *
+     * @return True when the result has been set, false when not.
+     */
+    @Override
+    public boolean hasResult()
+    {
+        return getResult() != null;
+    }
+
+    /**
+     * Returns the parent of this request.
+     * If this is set it means that this request is part of request chain.
+     *
+     * @return The parent of this request, or null if it has no parent.
+     */
+    @Nullable
+    @Override
+    @SuppressWarnings("unchecked")
+    public <T extends IToken> T getParent()
+    {
+        return (T) parent;
+    }
+
+    /**
+     * Method used to set the parent of a request.
+     *
+     * @param parent The new parent, or null to clear the existing one.
+     */
+    @Override
+    public <T extends IToken> void setParent(@Nullable T parent)
+    {
+        this.parent = parent;
+    }
+
+    /**
+     * Returns true if this request has a parent, false if not.
+     *
+     * @return true if this request has a parent, false if not.
+     */
+    @Override
+    public boolean hasParent()
+    {
+        return parent != null;
+    }
+
+    /**
+     * Method used to add a single Child.
+     *
+     * @param child The new child request to add.
+     */
+    @Override
+    public <T extends IToken> void addChild(@NotNull T child)
+    {
+        this.children.add(child);
+        Log.getLogger().debug("Added child:" + child + " to: " + getToken());
+    }
+
+    /**
+     * Method to add multiple children in a single call.
+     *
+     * @param children An array of children to add.
+     */
+    @Override
+    public <T extends IToken> void addChildren(@NotNull T... children)
+    {
+        for (IToken token : children)
+        {
+            addChild(token);
+        }
+    }
+
+    /**
+     * Method to add multiple children in a single call.
+     *
+     * @param children A collection of children to add.
+     */
+    @Override
+    public <T extends IToken> void addChildren(@NotNull Collection<T> children)
+    {
+        for (IToken token : children)
```

Comment:
![MAJOR](https://sonarsource.github.io/sonar-github/severity-major.png 'Severity: MAJOR') Rename "token" which hides the field declared at line 27. [![rule](https://sonarsource.github.io/sonar-github/rule.png)](https://sonar.minecolonies.com/coding_rules#rule_key=squid%3AHiddenFieldCheck)